File: LP-1/Assembler (Python)/Pass-2.py
from re import compile

tab = "\t"
newline = "\n"

pattern = compile(r'\(\'([A-Z]{,2})\',\s(\d+)\)')
# captures upto 2 consecutive uppercase alphabets and one or more digits sep by a comma within () eg: (AD,01)
# pattern is now compiled into an object that can be used to search for the above regex

class Pass2:
    def __init__(self):
        self.symTab = {}
        self.litTab = {}

        # files
        self.ICFile = open("intermediateCode.txt","r")
        self.symTabFile = open("symbolTable.txt","r")
        self.litTabFile = open("literalTable.txt","r")
        self.outputFile = open("machineCode.txt", "w")


    def convert(self, string):
        string = str(string)
        if len(string)==1:
            return "00" + string
        elif len(string)==2:
            return "0" + string
        elif len(string)==3:
            return string
        

    def readST(self):
        print("\n--------------------")
        print("SYMBOL TABLE")
        print("--------------------")
        for line in self.symTabFile.readlines():
            line = line.split("\t")
            index = int(line[0])
            location = int(line[2])
            self.symTab[index] = location
            print(index, location, sep=tab)
        print()

    
    def readLT(self):
        print("\n--------------------")
        print("LITERAL TABLE")
        print("--------------------")
        for line in self.litTabFile.readlines():
            line = line.split("\t")
            index = int(line[0])
            location = int(line[2])
            self.litTab[index] = location
            print(index, location, sep="\t")
        print()

    
    def generate(self):
        self.readST()
        self.readLT()
        print("\n-----------------------------")
        print("MACHINE CODE")
        print("-----------------------------")
        print("| LC | OPCODE | OP-1 | OP-2 |")
        print("-----------------------------")
        
        for line in self.ICFile.readlines():
            line = line.strip("\n")
            line = line.split("\t")

            find = pattern.search(line[0])  # if found, 'find' will hold the matched object, else None

            # group(1) will access captured group ([A-Z]{,2}) and group(2) will access (d+)
            if find.group(1)=="IS" or find.group(1)=="DL":
                machineCode = ""
                location = line[-2]
                machineCode += location + tab

                if find.group(1)=="IS":
                    machineCode += self.convert(find.group(2)) + tab

                    if find.group(2)=="10" or find.group(2)=="9":
                        find = pattern.search(line[1])
                        key = int(find.group(2))
                        machineCode += "000" + tab + self.convert(self.symTab[key]) + newline
                    elif find.group(2)=="0":
                        machineCode += "000" + tab + "000" + newline
                    else:
                        find = pattern.search(line[1])
                        machineCode += self.convert(find.group(2)) + tab
                        find = pattern.search(line[2])
                        if find.group(1)=="S":  
                            key = int(find.group(2))
                            machineCode += self.convert(self.symTab[key]) + newline
                        elif find.group(1)=="L":
                            key = int(find.group(2))
                            machineCode += self.convert(self.litTab[key]) + newline
                else:
                    if find.group(2)=="1":
                        machineCode += "000" + tab + "000" + tab
                        find = pattern.search(line[1])
                        machineCode += self.convert(find.group(2)) + newline
                    else:
                        machineCode += "000" + tab + "000" + tab + "000" + newline
            
            else:
                continue
            print(machineCode, end="")
            self.outputFile.write(machineCode)

        self.outputFile.close()
        self.symTabFile.close()
        self.litTabFile.close()


obj = Pass2()
obj.generate()

File: LP-1/Assembler (Java)/Pass2.java
import java.io.*;
import java.util.HashMap;
import java.util.Map;
import java.util.regex.*;

public class Pass2 {
    private Map<Integer, Integer> symTab = new HashMap<>();
    private Map<Integer, Integer> litTab = new HashMap<>();

    private BufferedReader ICFile;
    private BufferedReader symTabFile;
    private BufferedReader litTabFile;
    private BufferedWriter outputFile;

    private Pattern pattern = Pattern.compile("\\('\\([A-Z]{,2}\\)',\\s(\\d+)\\)");

    private static String tab = "\t";
    private static String newline = "\n";

    public Pass2() throws IOException {
        ICFile = new BufferedReader(new FileReader("intermediateCode.txt"));
        symTabFile = new BufferedReader(new FileReader("symbolTable.txt"));
        litTabFile = new BufferedReader(new FileReader("literalTable.txt"));
        outputFile = new BufferedWriter(new FileWriter("machineCode.txt"));
    }

    private String convert(int number) {
        String string = Integer.toString(number);
        if (string.length() == 1) {
            return "00" + string;
        } else if (string.length() == 2) {
            return "0" + string;
        } else {
            return string;
        }
    }

    private void readST() throws IOException {
        System.out.println("\n--------------------");
        System.out.println("SYMBOL TABLE");
        System.out.println("--------------------");
        String line;
        while ((line = symTabFile.readLine()) != null) {
            String[] parts = line.split("\t");
            int index = Integer.parseInt(parts[0]);
            int location = Integer.parseInt(parts[2]);
            symTab.put(index, location);
            System.out.println(index + tab + location);
        }
        System.out.println();
    }

    private void readLT() throws IOException {
        System.out.println("\n--------------------");
        System.out.println("LITERAL TABLE");
        System.out.println("--------------------");
        String line;
        while ((line = litTabFile.readLine()) != null) {
            String[] parts = line.split("\t");
            int index = Integer.parseInt(parts[0]);
            int location = Integer.parseInt(parts[2]);
            litTab.put(index, location);
            System.out.println(index + tab + location);
        }
        System.out.println();
    }

    private void generate() throws IOException {
        readST();
        readLT();
        System.out.println("\n----------------------------------------------------");
        System.out.println("MACHINE CODE");
        System.out.println("----------------------------------------------------");
        System.out.println("LC\tOPCODE\tOPERAND-1\tOPERAND-2");

        String line;
        while ((line = ICFile.readLine()) != null) {
            line = line.trim();
            String[] parts = line.split("\t");

            Matcher matcher = pattern.matcher(parts[0]);
            if (matcher.matches()) {
                String group1 = matcher.group(1);
                String group2 = matcher.group(2);

                String machineCode = "";
                String location = parts[parts.length - 2];
                machineCode += location + tab;

                if ("IS".equals(group1) || "DL".equals(group1)) {
                    machineCode += convert(Integer.parseInt(group2)) + tab;

                    if ("IS".equals(group1) && ("10".equals(group2) || "9".equals(group2))) {
                        matcher = pattern.matcher(parts[1]);
                        if (matcher.matches()) {
                            int key = Integer.parseInt(matcher.group(2));
                            machineCode += "000" + tab + convert(symTab.get(key)) + newline;
                        }
                    } else if ("0".equals(group2)) {
                        machineCode += "000" + tab + "000" + newline;
                    } else {
                        matcher = pattern.matcher(parts[1]);
                        if (matcher.matches()) {
                            machineCode += convert(Integer.parseInt(matcher.group(2))) + tab;

                            matcher = pattern.matcher(parts[2]);
                            if (matcher.matches()) {
                                if ("S".equals(matcher.group(1))) {
                                    int key = Integer.parseInt(matcher.group(2));
                                    machineCode += convert(symTab.get(key)) + newline;
                                } else if ("L".equals(matcher.group(1))) {
                                    int key = Integer.parseInt(matcher.group(2));
                                    machineCode += convert(litTab.get(key)) + newline;
                                }
                            }
                        }
                    }
                } else {
                    if ("1".equals(group2)) {
                        machineCode += "000" + tab + "000" + tab;

                        matcher = pattern.matcher(parts[1]);
                        if (matcher.matches()) {
                            machineCode += convert(Integer.parseInt(matcher.group(2))) + newline;
                        }
                    } else {
                        machineCode += "000" + tab + "000" + tab + "000" + newline;
                    }
                }

                System.out.print(machineCode);
                outputFile.write(machineCode);
            }
        }

        outputFile.close();
        symTabFile.close();
        litTabFile.close();
    }

    public static void main(String[] args) throws IOException {
        Pass2 obj = new Pass2();
        obj.generate();
    }
}
